Sort check on JSON

kivikakk opened the pull request:
<!-- Please describe your changes on the following line: -->
Check that JSON keys are ordered, and that there's no duplicates (for `resources/prefs.json` and `resources/package-prefs.json`).

---
<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `__` with appropriate data: -->
- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [x]  🆕 `./mach test-tidy --self-test` does not report any errors
- [x] These changes fix #12283

<!-- Either: -->
- [x] There are tests for these changes 🎉

<!-- Pull requests that do not address these steps are welcome, but they will require additional verification as part of the review process. -->

<!-- Reviewable:start -->

---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/14092)
<!-- Reviewable:end -->

kivikakk commented:
(Note that this builds directly on #13789!)

kivikakk commented:
Oh, missed `./mach test-tidy --no-progress --all`.

kivikakk commented:
I note this applied to `etc/ci/performance/firefox/addon/package.json` as well; I've reordered the `fennec` and `firefox` keys accordingly, but I'm not sure if this should be occurring here or not.

kivikakk commented:
Haha, and now it's complaining about a `package.json`. This doesn't seem quite right.

emilio commented:
r? @Wafflespeanut

kivikakk commented:
Took a different tack, and am only applying the alphabetic sorting rule to `./resources/**`.

Wafflespeanut commented on python/tidy/servo_tidy/tidy.py:
```diff
@@ -631,28 +631,55 @@ def check_webidl_spec(file_name, contents):
     for key in keys:
         if key in seen_keys:
-            raise KeyError(key)
+            raise DuplicateKeyError(key)
```
Do we really need two error objects just for the sake of raising key errors. Since we're catching the errors later anyway, we could just

``` python
raise KeyError("reason")
```

This way, we also won't be needing any additional attributes later and `yield` the reason directly.

kivikakk replied:
Sure, I'm happy with that. I like to be a little declarative, but this is so small …

Wafflespeanut commented on python/tidy/servo_tidy/tidy.py:
```diff
@@ -631,28 +631,55 @@ def check_webidl_spec(file_name, contents):
+def check_json_requirements(filename):
+    def check_fn(key_value_pairs):
+        check_for_possible_duplicate_json_keys(key_value_pairs)
+        if filename.startswith("./resources/"):
```
Having something like this would be specific to Servo, which we don't want, because tidy is published every once in a while, and is being used by other projects. We have a config file in the root for this (`servo-tidy.toml`). But, I doubt this would be necessary. We can check for sorting and duplicated keys at the same time.

Wafflespeanut replied:
I now think of having a key in the config file, say `check_duplicate_json_keys` which has a list of file paths. We could have the paths to JSONs in `resources` in that file, and get it from the config here.

kivikakk replied:
(Tomorrow, though; I'm heading to sleep now!)

Wafflespeanut replied:
Sorry, I meant `check_ordered_json_keys`, because we need the ordering only for the json files in `resources/` (right now). Duplicate keys check could be done for all jsons, so that's fine :smile:

kivikakk replied:
Good point! :) I'll do that tonight.

Wafflespeanut commented on python/tidy/servo_tidy/tidy.py:
```diff
@@ -631,28 +631,55 @@ def check_webidl_spec(file_name, contents):
+    except DuplicateKeyError as e:
+        yield (None, "Duplicated Key (%s)" % e.key)
+    except UnsortedKeyError as e:
+        yield (None, "Unordered key (found %s before %s)" % (e.a, e.b))
```
If we just `raise KeyError(...)` there, we could just do something like,

``` python
except KeyError as e:
    yield (None, e)
```

kivikakk replied:
Indeed! I figured it's neater to have the message generation in one place and leave the functions to be a bit more declarative or pure, but this is fine too.

Wafflespeanut commented on python/tidy/servo_tidy/tidy.py:
```diff
@@ -631,28 +631,55 @@ def check_webidl_spec(file_name, contents):
 
 
+def check_for_alphabetical_sorted_json_keys(key_value_pairs):
+    for a, b in zip(key_value_pairs[:-1], key_value_pairs[1:]):
```
Sweet!

kivikakk commented:
It wasn't actually on top of #13789 exactly; I had squashed it down to represent prior work — it was more of an FYI. I've squashed that away entirely now; I think this is back to you! 😺

kivikakk commented:
@Wafflespeanut back to you! :)

Wafflespeanut commented:
LGTM! Can you please squash the commits appropriately?

Commit:
Check for JSON key non-duplication and order

Commit:
Fix existing pref JSONs

kivikakk commented:
@Wafflespeanut: done, thanks! 😸

Wafflespeanut commented:
Thanks for doing this work! :)

@bors-servo r+

bors-servo commented:
:pushpin: Commit 4fe1658 has been approved by `Wafflespeanut`

<!-- @bors-servo r=Wafflespeanut 4fe165846598d7999224833396547a96fd4675fb -->

bors-servo commented:
:hourglass: Testing commit 4fe1658 with merge 94d0c48...

bors-servo commented:
:sunny: Test successful - [arm32](http://build.servo.org/builders/arm32/builds/3943), [arm64](http://build.servo.org/builders/arm64/builds/3949), [linux-dev](http://build.servo.org/builders/linux-dev/builds/4042), [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/685), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/684), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/4027), [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/4023), [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/707), [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/720), [windows-dev](http://build.servo.org/builders/windows-dev/builds/1955)
